enhancements/network/sdn-live-migration.md

```diff
@@ -0,0 +1,476 @@
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+
+## Summary
+Migrating the CNI network provider network of a running cluster from
+OpenShift SDN to OVN Kubernetes without service interruption. During the
+migration, we will partition the cluster into two sets of nodes controlled by
+different network plugins. We will utilize the Hybrid overlay feature of
+OVN Kubernetes to connect the networks of the two CNI network plugins. So that
+pods on each side can still talk to pods on the other side.
+
+## Motivation
+
+For some Openshift users, they have very high requirements on service
+availability. The current SDN migration solution, which will cause a service
+interruption, is not acceptable.
+
+### Goals
+
+- Migrate the cluster network provider from OpenShift SDN to OVN Kubernetes for
+  an existing cluster.
+  - The OVN Kubernetes will be running in the IC multi-zone mode.
+- This solution will work on all platforms managed by the SD team.
+  - ARO (Azure Red Hat OpenShift)
+  - OSDv4 (OpenShift Dedicated) on GCP
+  - OSDv4 (OpenShift Dedicated) on AWS
+  - ROSA Classic (Red Hat OpenShift Service on AWS)
+- This is an in-place migration without requiring extra nodes.
+- The impact on workload of the migration will be similar to that of an OCP
+  upgrade.
+- The solution can work at scale, e.g., in a large cluster with 100+ nodes.
+- The migration operation will be able to be rolled back if needed.
+
+### Non-Goals
+
+- Support for migration to other network providers
+- The necessary GUI change in Openshift Cluster Manager
+- Ensure the following features remain working during a live migration:
+  - Egress IP
+  - Egress Router
+  - Multicast
+- Migration from SDN Multi-tenant mode
+- Support Hypershift clusters
+
+## Pre-requisites
+
+- OVN-IC multi-zone mode in OCP is dev-complete.
+
+## Proposal
+
+The key problem with doing a live SDN migration is that we need to maintain the
+connectivity of the cluster network during the migration when pods are attached
+to different networks. We propose to utilize the OVN Kubernetes hybrid overlay
+feature to connect the networks owned by OpenShift SDN and OVN Kubernetes.
+
+- We will run different plugins on different nodes, but both plugins will know
+  how to reach pods owned by the other plugin, so all pods, services, etc.
+  remain connected.
+- During migration, CNO will take original-plugin nodes one by one and convert
+  them to destination-plugin nodes, rebooting them in the process.
+- The cluster network CIDR will remain unchanged, as will the node host subnet
+  of each node.
+- NetworkPolicy will work correctly throughout the migration.
+
+### Limitations
+
+- The following features, which are only supported by OpenShift SDN but not by
+  OVN Kubernetes, will stop working when the migration is started.
+  - Multitenant Isolation
+- The Egress Router feature is supported by both OpenShift SDN and OVN
```

Comment:
I am not sure if there metrics/telemetry available. @martinkennelly may know better.

Those CRs/CRDs will be removed after the migration. During the migration, it will be there but have no function. For egressIP, egressFW, and multicast. we will convert the CRs to new ones that are used by the target CNI.

Reply:
That metric works for OVN-K and SDN.

Reply:
Thank you, these are helpful. I can see us setting up some monitoring/alerting around these metrics (we have the ability to push custom telemetry to/from the managed fleet) in order to determine usage of these features and warn users ahead of the migration.

Reply:
I am creating a doc to give to you folks that will have this: https://issues.redhat.com/browse/SDN-4178
Id appreciate any feedback. Please watch this story.